### ISSUE TITLE
ExpansionPoint

### ISSUE THREAD

**Issue by mohammad7613** [ilabcode/pyhgf#406]
# Question: nonlinear value-coupling derivatives are evaluated at `mean` rather than `expected_mean`

## Summary

While checking the continuous-node posterior updates for nonlinear value coupling, I noticed that the coupling-function derivatives are evaluated at the node's current/previous posterior mean, `mean`, while the posterior correction is applied around the predicted mean, `expected_mean`.

On current `master`, the Python and Rust backends appear to agree on the value-coupling precision formula, including the single factor of the value-coupling strength on the second-derivative term. My remaining question is specifically about the Taylor/Laplace expansion point used for nonlinear value-coupling derivatives.

This is not only an implementation detail: the current implementation appears to follow the generalized HGF paper's equations 50 and 51, which also evaluate the nonlinear derivatives at the previous posterior mean. My question is therefore whether this expansion point is intentional in the generalized value-coupling setting, or whether the previous-posterior expansion point was inherited from the classical HGF case where it coincides with the predicted mean.

Current upstream checked:

- Repository: `https://github.com/ComputationalPsychiatry/pyhgf`
- Current `master`: `6806ff728e8318965020bf3520c53e51b3524f47`
- Commit message: `Add the Transformer parity test suite (#404)`
- Older reference where I first noticed related issues: `726437488379715fe5ecb13fde2991cbafa05fd5`, `refactor and simplify the updates for volatile state nodes (#354)`

## Current behaviour

For a node `b` with a nonlinear value-coupling function `g` to a child `a`, the Python backend evaluates derivatives at:

```python
attributes[node_idx]["mean"]
```

In the mean update, this appears in:

```python
coupling_fn_prime = grad(coupling_fn)(attributes[node_idx]["mean"])
```

In the precision update, this appears in the value-coupling precision-weighted prediction error helper:

```python
grad(coupling_fn)(attributes[node_idx]["mean"])
grad(grad(coupling_fn))(attributes[node_idx]["mean"])
```

The Rust backend follows the same expansion point:

```rust
let parent_mean = network.attributes.states[node_idx].mean;
let g_prime = (cf.df)(parent_mean);
let g_second = (cf.d2f)(parent_mean);
```

The posterior mean itself is then updated as a correction to the predicted mean:

```python
posterior_mean = attributes[node_idx]["expected_mean"] + ...
```

So the update uses derivatives evaluated at `mean`, but applies the local quadratic/Gaussian correction around `expected_mean`.

## Relation to the GHGF paper

The same expansion-point choice appears in the generalized HGF paper. In the nonlinear value-coupling section, equations 50 and 51 evaluate the derivative terms at:

```text
mu_b^(k-1)
```

while the mean update is written as a correction to the predicted mean:

```text
mu_b^(k) = mu_hat_b^(k) + correction
```

The surrounding text also states that node `b` needs access to its own previous posterior mean for the update. In the appendix, the derivation explicitly says that the quadratic approximation follows Mathys et al. (2011) and uses the expansion point:

```text
mu_b^(k-1)
```

So the current implementation seems faithful to the paper. The question is whether the paper's expansion point remains self-consistent after generalizing the HGF to nodes whose predicted mean can differ from the previous posterior mean.

## Why I think this may be inconsistent

In my derivation, the local variational objective for node `b` at time `t` is expanded around the prediction before the current observation is assimilated:

```text
x_b* = expected_mean_b,t = mu_hat_b,t
```

For a nonlinear value child contribution of the form:

```text
T(x_b) = -1/2 * pi_a * (mu_a,t - m_a(x_b))^2
```

with:

```text
m_a(x_b) = ... + kappa * g(x_b)
```

the local Gaussian/Laplace update gives:

```text
posterior_mean_b = expected_mean_b + posterior_variance_b * I'(expected_mean_b)
posterior_precision_b = -I''(expected_mean_b)
```

Under this reading, both `g'` and `g''` should be evaluated at:

```python
attributes[node_idx]["expected_mean"]
```

rather than:

```python
attributes[node_idx]["mean"]
```

The two expansion points coincide when the node has no value parents, because then:

```text
expected_mean_b,t = mean_b,t-1
```

But for a node that itself has value parents, `expected_mean` can differ from `mean`. In that case, nonlinear coupling derivatives can differ substantially, especially for saturating nonlinearities such as `tanh`. 

## Possible justification for the current expansion point

One possible justification is historical and computational. The paper states that the nonlinear update follows the Mathys et al. (2011) HGF update rules, which use a quadratic approximation around the previous posterior belief:

```text
mu_b^(k-1)
```

That is natural in the classical HGF random-walk setting, because for a value parent without value parents of its own:

```text
mu_hat_b^(k) = mu_b^(k-1)
```

In that case, evaluating derivatives at `mean` and evaluating them at `expected_mean` are the same operation. This also explains why the issue is invisible in standard linear-coupling or simple chain examples.

However, in the generalized HGF, a node can itself have value parents, tonic drift, or non-unit autoconnection. Then:

```text
mu_hat_b^(k) != mu_b^(k-1)
```

If the expansion point is truly `mu_b^(k-1)`, a complete local Gaussian update should retain the prior-gradient residual caused by expanding the prior term away from its center `mu_hat_b^(k)`. If the update is instead written as:

```text
mu_b^(k) = mu_hat_b^(k) + correction
```

then the self-consistent single-point Laplace expansion seems to require evaluating the likelihood/coupling derivatives at:

```text
mu_hat_b^(k)
```

This is the point I would like to clarify.

## Why this matters

This affects models with custom nonlinear `coupling_fn` on value-coupling edges. Linear coupling is unaffected because:

```text
g'(x) = 1
g''(x) = 0
```

I tested a small diagnostic setup with:

```text
x2 -> x1 through tanh(x2)
x1 -> binary observation
```

When `x2` is forced to pass through the nonlinear sensitive range of `tanh`, changing the derivative expansion point from `mean` to `expected_mean` changes the inferred hidden-state trajectories and posterior precisions. In one frozen diagnostic dataset where `x2` ranged approximately from `-11.37` to `+9.69`, the `expected_mean` expansion point slightly improved both hidden-state recovery and binary predictive log-loss:

```text
Mixed/current expansion point:
x1 RMSE = 0.1679
x2 RMSE = 0.9074
pred_logloss = 0.3657

Expected-mean expansion point:
x1 RMSE = 0.1648
x2 RMSE = 0.8932
pred_logloss = 0.3653
```

This is not meant as a general benchmark claim, only as evidence that the expansion-point choice is observable in inference, not just a cosmetic implementation detail.

## What I am asking

Is evaluating nonlinear value-coupling derivatives at `mean` intentional in the generalized value-coupling setting, or should these derivatives be evaluated at `expected_mean` to match the local expansion point used by the posterior update?

If `mean` is intentional, I would appreciate clarification of the intended derivation, especially how the prior-gradient residual is handled when `mean != expected_mean`. If `expected_mean` is the intended expansion point, I can open a PR that changes the Python and Rust implementations consistently and adds a regression test using nonlinear value coupling with a parent node for which:

```text
mean != expected_mean
```

## Implementation locations

The relevant locations on current `master` at commit `6806ff728e8318965020bf3520c53e51b3524f47` are:

### Python backend

Mean update:

- `pyhgf/updates/posterior/continuous/posterior_update_mean_continuous_node.py`
- Standard update: line 120 evaluates `g'` at `attributes[node_idx]["mean"]`.
- Standard posterior mean: lines 189-190 add the correction to `attributes[node_idx]["expected_mean"]`.
- Mean-field update: line 322 evaluates `g'` at `attributes[node_idx]["mean"]`.
- Mean-field posterior mean: lines 355-356 add the correction to `attributes[node_idx]["expected_mean"]`.

Stable links:

- https://github.com/ComputationalPsychiatry/pyhgf/blob/6806ff728e8318965020bf3520c53e51b3524f47/pyhgf/updates/posterior/continuous/posterior_update_mean_continuous_node.py#L120
- https://github.com/ComputationalPsychiatry/pyhgf/blob/6806ff728e8318965020bf3520c53e51b3524f47/pyhgf/updates/posterior/continuous/posterior_update_mean_continuous_node.py#L189
- https://github.com/ComputationalPsychiatry/pyhgf/blob/6806ff728e8318965020bf3520c53e51b3524f47/pyhgf/updates/posterior/continuous/posterior_update_mean_continuous_node.py#L322
- https://github.com/ComputationalPsychiatry/pyhgf/blob/6806ff728e8318965020bf3520c53e51b3524f47/pyhgf/updates/posterior/continuous/posterior_update_mean_continuous_node.py#L355

Precision update:

- `pyhgf/updates/posterior/continuous/posterior_update_precision_continuous_node.py`
- Shared value-coupling precision helper: `_value_coupling_pwpe`, starting at line 83.
- Line 117 evaluates `g'` at `attributes[node_idx]["mean"]`.
- Line 124 evaluates `g''` at `attributes[node_idx]["mean"]`.
- This helper is used by standard precision update, mean-field precision update, eHGF precision update, and eHGF mean-field precision update.

Stable links:

- https://github.com/ComputationalPsychiatry/pyhgf/blob/6806ff728e8318965020bf3520c53e51b3524f47/pyhgf/updates/posterior/continuous/posterior_update_precision_continuous_node.py#L83
- https://github.com/ComputationalPsychiatry/pyhgf/blob/6806ff728e8318965020bf3520c53e51b3524f47/pyhgf/updates/posterior/continuous/posterior_update_precision_continuous_node.py#L117
- https://github.com/ComputationalPsychiatry/pyhgf/blob/6806ff728e8318965020bf3520c53e51b3524f47/pyhgf/updates/posterior/continuous/posterior_update_precision_continuous_node.py#L124
- https://github.com/ComputationalPsychiatry/pyhgf/blob/6806ff728e8318965020bf3520c53e51b3524f47/pyhgf/updates/posterior/continuous/posterior_update_precision_continuous_node.py#L317
- https://github.com/ComputationalPsychiatry/pyhgf/blob/6806ff728e8318965020bf3520c53e51b3524f47/pyhgf/updates/posterior/continuous/posterior_update_precision_continuous_node.py#L426
- https://github.com/ComputationalPsychiatry/pyhgf/blob/6806ff728e8318965020bf3520c53e51b3524f47/pyhgf/updates/posterior/continuous/posterior_update_precision_continuous_node.py#L628
- https://github.com/ComputationalPsychiatry/pyhgf/blob/6806ff728e8318965020bf3520c53e51b3524f47/pyhgf/updates/posterior/continuous/posterior_update_precision_continuous_node.py#L659

### Rust backend

Continuous posterior updates:

- `src/updates/posterior/continuous.rs`
- Standard precision helper: line 50 sets `parent_mean = network.attributes.states[node_idx].mean`; lines 61-62 evaluate `g'` and `g''` at that point.
- Standard mean helper: line 141 sets `parent_mean = ...mean`; line 151 evaluates `g'` at that point.
- Mean-field precision helper: line 366 sets `parent_mean = ...mean`; lines 377-378 evaluate `g'` and `g''`.
- eHGF precision helper: line 450 sets `parent_mean = ...mean`; lines 461-462 evaluate `g'` and `g''`.
- eHGF mean-field precision helper: line 516 sets `parent_mean = ...mean`; lines 527-528 evaluate `g'` and `g''`.
- Mean-field mean helper: line 570 sets `parent_mean = ...mean`; line 580 evaluates `g'`.
- The posterior means are still written as corrections to `expected_mean`, for example lines 208 and 217 in the standard update path.

Stable links:

- https://github.com/ComputationalPsychiatry/pyhgf/blob/6806ff728e8318965020bf3520c53e51b3524f47/src/updates/posterior/continuous.rs#L50
- https://github.com/ComputationalPsychiatry/pyhgf/blob/6806ff728e8318965020bf3520c53e51b3524f47/src/updates/posterior/continuous.rs#L61
- https://github.com/ComputationalPsychiatry/pyhgf/blob/6806ff728e8318965020bf3520c53e51b3524f47/src/updates/posterior/continuous.rs#L141
- https://github.com/ComputationalPsychiatry/pyhgf/blob/6806ff728e8318965020bf3520c53e51b3524f47/src/updates/posterior/continuous.rs#L151
- https://github.com/ComputationalPsychiatry/pyhgf/blob/6806ff728e8318965020bf3520c53e51b3524f47/src/updates/posterior/continuous.rs#L366
- https://github.com/ComputationalPsychiatry/pyhgf/blob/6806ff728e8318965020bf3520c53e51b3524f47/src/updates/posterior/continuous.rs#L450
- https://github.com/ComputationalPsychiatry/pyhgf/blob/6806ff728e8318965020bf3520c53e51b3524f47/src/updates/posterior/continuous.rs#L516
- https://github.com/ComputationalPsychiatry/pyhgf/blob/6806ff728e8318965020bf3520c53e51b3524f47/src/updates/posterior/continuous.rs#L570
- https://github.com/ComputationalPsychiatry/pyhgf/blob/6806ff728e8318965020bf3520c53e51b3524f47/src/updates/posterior/continuous.rs#L217

## Notes on older observations

I originally drafted a broader issue against `master` at `726437488379715fe5ecb13fde2991cbafa05fd5`, where I observed a Python/Rust mismatch in the nonlinear value-coupling precision update and a missing `kappa` factor on the second-derivative term.

Those points appear resolved on current `master` at `6806ff728e8318965020bf3520c53e51b3524f47`. The current issue is therefore narrowed to the expansion-point question only.
